[GPU Process] [FormControls] Add a ControlPart for ToggleButton

shallawa opened the pull request:
#### a5f12d1b9232202fd842ed8b615d2baed86aa271
<pre>
[GPU Process] [FormControls] Add a ControlPart for ToggleButton
<a href="https://bugs.webkit.org/show_bug.cgi?id=249638">https://bugs.webkit.org/show_bug.cgi?id=249638</a>
rdar://103548470

Reviewed by Aditya Keerthi.

It will handle drawing the checkbox and the radio buttons. ButtonControlMac will
be introduced as the base class of ToggleButtonMac. It will also be the base class
of the ButtonMac which will be submitted in a future patch.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ToggleButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm: Added.
(WebCore::ButtonControlMac::ButtonControlMac):
(WebCore::ButtonControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::createToggleButtonCell):
(WebCore::ControlFactoryMac::checkboxCell const):
(WebCore::ControlFactoryMac::radioCell const):
(WebCore::ControlFactoryMac::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
(WebCore::ControlMac::cellSize const):
(WebCore::ControlMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::calculateControlSize const):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm: Added.
(WebCore::ToggleButtonMac::ToggleButtonMac):
(WebCore::ToggleButtonMac::cellSize const):
(WebCore::ToggleButtonMac::cellOutsets const):
(WebCore::ToggleButtonMac::draw):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::paint):
(WebCore::paintToggleButton): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::canCreateControlPartForBorderOnly const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/258177@main">https://commits.webkit.org/258177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/68b4a7a201f1def90476faa9fe87e4df251f9a04

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101083 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10237 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34134 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110381 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170636 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105069 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11168 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1112 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93496 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108213 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106865 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8458 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/91725 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35063 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23122 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78055 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3899 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24639 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3931 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1043 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10041 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44147 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5610 "Found 1 new test failure: media/video-playback-quality.html (failure)") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5696 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2947 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.h:
```diff
@@ -0,0 +1,46 @@
+
+class ButtonControlMac : public ControlMac {
+public:
+    ButtonControlMac(ControlPart&, ControlFactoryMac&, NSButtonCell*);
```
Nit: `NSButtonCell *` since it's a Objective-C type.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
```diff
@@ -0,0 +1,48 @@
+
+class ToggleButtonMac final : public ButtonControlMac {
+public:
+    ToggleButtonMac(ToggleButtonPart& owningPart, ControlFactoryMac&, NSButtonCell*);
```
Nit: `NSButtonCell *`.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h:
```diff
@@ -0,0 +1,48 @@
+
+private:
+    IntSize cellSize(NSControlSize, const ControlStyle&) const override;
+    IntOutsets cellOutsets(NSControlSize, const ControlStyle&) const override;
```
`final`?

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm:
```diff
@@ -0,0 +1,147 @@
+
+namespace WebCore {
+
+ToggleButtonMac::ToggleButtonMac(ToggleButtonPart& owningPart, ControlFactoryMac& controlFactory, NSButtonCell* buttonCell)
```
Nit: `NSButtonCell *`.

shallawa replied:
Fixed.

pxlcoder commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -408,74 +408,6 @@ static bool drawCellFocusRing(NSCell *cell, NSRect cellFrame, NSView *controlVie
-    } else {
-        if (!toggleButtonCell) {
-            if (buttonType == ControlPartType::Checkbox)
-                toggleButtonCell = sharedCheckboxCell(controlStates, zoomedRectSize, zoomFactor);
```
We should think about how to implement `inflateControlPaintRect` in terms of the new control architecture, so that we can share more code, and remove `sharedCheckboxCell`, `sharedRadioCell`, and related methods from `ThemeMac`.

Could also be done in a separate patch, once all buttons are ready.

shallawa committed:
[GPU Process] [FormControls] Add a ControlPart for ToggleButton
https://bugs.webkit.org/show_bug.cgi?id=249638
rdar://103548470

Reviewed by Aditya Keerthi.

It will handle drawing the checkbox and the radio buttons. ButtonControlMac will
be introduced as the base class of ToggleButtonMac. It will also be the base class
of the ButtonMac which will be submitted in a future patch.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/ToggleButtonPart.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonControlMac.mm: Added.
(WebCore::ButtonControlMac::ButtonControlMac):
(WebCore::ButtonControlMac::updateCellStates):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::createToggleButtonCell):
(WebCore::ControlFactoryMac::checkboxCell const):
(WebCore::ControlFactoryMac::radioCell const):
(WebCore::ControlFactoryMac::createPlatformToggleButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
(WebCore::ControlMac::cellSize const):
(WebCore::ControlMac::cellOutsets const):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::calculateControlSize const):
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.h: Copied from Source/WebCore/platform/graphics/controls/ControlFactory.h.
* Source/WebCore/platform/graphics/mac/controls/ToggleButtonMac.mm: Added.
(WebCore::ToggleButtonMac::ToggleButtonMac):
(WebCore::ToggleButtonMac::cellSize const):
(WebCore::ToggleButtonMac::cellOutsets const):
(WebCore::ToggleButtonMac::draw):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::ThemeMac::paint):
(WebCore::paintToggleButton): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
(WebCore::RenderThemeMac::canCreateControlPartForBorderOnly const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):

Canonical link: https://commits.webkit.org/258177@main
